### PR TITLE
Github Action workflow for ropenms

### PR DESCRIPTION
**Pull request by 24sharkS**
# Description
- added workflow `ropenms-build.yml` for testing the build of ropenms package.
- can be installed using the package tarball uploaded as artifact for any platform. (platform independent)
- continuation of work mentioned here (https://github.com/OpenMS/OpenMS/issues/4843#issuecomment-688730002)

# Checklist:
- [ ] Make sure that you are listed in the AUTHORS file
- [ ] Add relevant changes and new features to the CHANGELOG file
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] New and existing unit tests pass locally with my changes
- [ ] Updated or added python bindings for changed or new classes. (Tick if no updates were necessary.)

# How can I get additional information on failed tests during CI:
If your PR is failing you can check out 
- http://cdash.openms.de/index.php?project=OpenMS and look for your PR. If you click in the column that lists the failed tests you will get detailed error messages.

### CODE REVIEW

**Comment by oliveralka** [comment]
@24sharkS Nice work - thanks a lot! Could you please merge develop to be up to date and have a look at the style issues (https://github.com/OpenMS/OpenMS/pull/5226/checks?check_run_id=2093294196).

**Comment by 24sharkS** [comment]
> @24sharkS Nice work - thanks a lot! Could you please merge develop to be up to date and have a look at the style issues (https://github.com/OpenMS/OpenMS/pull/5226/checks?check_run_id=2093294196).

I did make the necessary style changes but CodeFactor still shows the old ones.

**Comment by Leon-Bichmann** [comment]
Hi Shekhar,
When trying to install it from `OpenMS/src/ropenms` I run into problems of python, R and pyopenms versions:
Does it require a specific version of these 3 dependencies and are those the same versions on all platforms (linux, macos, windows)?

For me (using macos), it requires pyopenms==2.5.0 for instance and in the `ropenms-build.yml` we saw that python versions are different on the platforms.

File: .github/workflows/ropenms-build.yml [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+    - uses: actions/checkout@v2
+      with:
+        path: OpenMS
+    
+    # Custom action installs the latest miniconda version and create an environment to manage python dependencies.
+    - uses: conda-incubator/setup-miniconda@v2
+      name: Miniconda Installation
+      with:
+        miniconda-version: "latest"
+        python-version: 3.8
```
Comment (Leon-Bichmann) [inline .github/workflows/ropenms-build.yml:38]: here  python==3.8

File: .github/workflows/ropenms-build.yml [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+    - uses: actions/checkout@v2
+      with:
+        path: OpenMS
+
+    # Custom action installs the latest miniconda version and create an environment to manage python dependencies.
+    - uses: conda-incubator/setup-miniconda@v2
+      name: Miniconda Installation
+      with:
+        miniconda-version: "latest"
+        python-version: 3.7
```
Comment (Leon-Bichmann) [inline .github/workflows/ropenms-build.yml:96]: here python==3.7

File: src/pyOpenMS/R-addons/AASequence.R [hunk 3]
```diff
@@ -0,0 +1,14 @@
+
+# void getAAFrequencies(Map[String, size_t]) nogil except + # wrap-ignore
+AASequence$$set("public","getAAFrequencies", function(map) {
```
Comment (tjeerdijk) [inline src/pyOpenMS/R-addons/AASequence.R:3]: What is the reason for the double dollar "$$" referencing. In the reticulate manuals, I only see single dollars "$" for this purpose.

File: src/pyOpenMS/R-addons/DPosition.R [hunk 4]
```diff
@@ -0,0 +1,78 @@
+
```
Comment (tjeerdijk) [inline src/pyOpenMS/R-addons/DPosition.R:1]: You provide some R-style documentation here but not in general. The standard roxygen style documentation is like this (from https://r-pkgs.org/man.html):
#' Add together two numbers
#' 
#' @param x A number.
#' @param y A number.
#' @return The sum of \code{x} and \code{y}.
#' @examples
#' add(1, 1)
#' add(10, 1)
add <- function(x, y) {
  x + y
}

File: .github/workflows/ropenms-build.yml (hunk 2 above)
Comment (24sharkS) [reply .github/workflows/ropenms-build.yml:96]: I have tested for version 3.7 but it should work with others too.
https://pyopenms.readthedocs.io/en/latest/installation.html#index-0:~:text=Note%20that%20for%20Windows%2C%20we%20only%20support%20Python%203.5%2C%203.6%20and%203.7)

**Comment by 24sharkS** [comment]
> Hi Shekhar,
> When trying to install it from `OpenMS/src/ropenms` I run into problems of python, R and pyopenms versions:
> Does it require a specific version of these 3 dependencies and are those the same versions on all platforms (linux, macos, windows)?
> 
> For me (using macos), it requires pyopenms==2.5.0 for instance and in the `ropenms-build.yml` we saw that python versions are different on the platforms.
> 
> Would it work with pyopenms==2.6.0 as well?

Hi @Leon-Bichmann, so the idea was to install the package using the tarball produced as artifact. As for `OpenMS/src/ropenms`, the documentation files and Namespace are added later while building the package. 
The generated bundled package is independent of R version and can be installed on any platform. Any other python version compatible with pyopenms for a specific platform should work.
I have tested the build with specified python versions but it should work with others as well. Any other version of R should also work (as long as the package dependencies are available for it).

File: src/ropenms/DESCRIPTION [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+Title: OpenMS R Wrapper
+Version: 1.0.0
+Author: Shekhar Shukla
+Maintainer: Shekhar <nk.shekhar24@gmail.com>
+Description: This package provides access to the classes of pyopenms
+	library(https://pyopenms.readthedocs.io/en/latest/#)
+Config/reticulate:
+  list(
+    packages = list(
+      list(package = "pyopenms", version = "2.5.0", pip = TRUE)
```
Comment (Leon-Bichmann) [inline src/ropenms/DESCRIPTION:12]: ok maybe take out the version of pyopenms here

Comment (Leon-Bichmann) [inline src/ropenms/DESCRIPTION:12]:
```suggestion
      list(package = "pyopenms", version >= "2.5.0", pip = TRUE)
```

File: .github/workflows/ropenms-build.yml (hunk 2 above)
Comment (Leon-Bichmann) [reply .github/workflows/ropenms-build.yml:96]: ok great

**Comment by oliveralka** [comment]
@Leon-Bichmann @24sharkS I am little bit out of the loop - sorry -, what is still missing here?

**Comment by Leon-Bichmann** [comment]
Hey, no worries - I think in the end it was all pretty much ready to go. Just some more comments in the code and maybe also in R style like @tjeerdijk suggested could help. Otherwise we tested it also manually and it worked fine.

**Comment by tjeerdijk** [comment]
Leon and I successfully installed and ran ropenms on a fresh install (of R, python and OpenMS). My next step would be installing it within my existing R/Rstudio installation. Due to Season of Docs stuff I have had no time for more ropenms testing yet …. sorry. (also wanted to wait for next release of R, version 4.1 is coming out this week)

**Comment by oliveralka** [comment]
rebuild jenkins

**Comment by oliveralka** [comment]
@tjeerdijk What is left to do here? Working on style fixes, but in general, this one is ready to merge, or?

**Comment by oliveralka** [comment]
The other CodeFactor issues are due to the use of python single tuples, which have a trailing ",". e.g. 

```
    def get_base_types(self):
        return ("DPosition2",)
```

I guess that should be fine.

**Comment by oliveralka** [comment]
I think if possible we should merge this for Release 2.7, to have automated R bindings for the most used classes.

**Comment by timosachsenberg** [comment]
- We changed quite a bit on the pyOpenMS side. Is this still up to date?
- Do the tests run with the newest pyOpenMS version?
- the PR is rather large. Would it make sense to move it into a submodule?

**Comment by oliveralka** [comment]
> * We changed quite a bit on the pyOpenMS side. Is this still up to date?
> * Do the tests run with the newest pyOpenMS version?
@tjeerdijk Would you be able to test this? 
> * the PR is rather large. Would it make sense to move it into a submodule?
I think that would make sense, but I do not have time for it and I am not sure how easy it is, since it is tightly integrated with pyopenms.

**Comment by timosachsenberg** [comment]
e.g., pyopenms_1.R is huge and part of the PR. Shouldn't that file be created automatically?

**Comment by tjeerdijk** [comment]
On 13. Jul 2021, at 14:45, Oliver Alka ***@***.***> wrote:
> 
> We changed quite a bit on the pyOpenMS side. Is this still up to date?
> Do the tests run with the newest pyOpenMS version?
> @tjeerdijk <https://github.com/tjeerdijk> Would you be able to test this?
Yes, my (default) plan for the upcoming hackathon (16 July) is to test this PR (unless there are more urgent matters)

> the PR is rather large. Would it make sense to move it into a submodule?
> I think that would make sense, but I do not have time for it and I am not sure how easy it is, since it is tightly integrated with pyopenms.
Splitting makes sense, can we discuss next Friday?

**Comment by oliveralka** [comment]
@tjeerdijk Where you able to have a look at this PR and test it?

**Comment by tjeerdijk** [comment]
Sorry, no I have not. I worked on the OpenMS web site and wikipedia page during the hackathon and now my time goes into pyOpenMS and will continue to do so while Rahul is still around until after my ASMS poster on the pyOpenMS-Binder integration.

So it is till on my to-do list but for late fall of 2021.

**Pull request closed without merging**